rsconnect/bundle.py

```diff
@@ -580,11 +580,11 @@ def make_quarto_source_bundle(
 
     base_dir = file_or_directory
     if not isdir(file_or_directory):
-        base_dir = basename(file_or_directory)
+        base_dir = dirname(file_or_directory)
 
     with tarfile.open(mode="w:gz", fileobj=bundle_file) as bundle:
         bundle_add_buffer(bundle, "manifest.json", json.dumps(manifest, indent=2))
-        if environment:
+        if environment and environment.source != 'file':
```

Comment:
This is extra protection for the `bundle_add_buffer` function to be called on in-memory `requirements.txt` file, not an actual file on disk.

Reply:
Regardless of whether the package requirements were sourced from a file or from `pip freeze`, the environment object should contain the correct set. We need to guarantee that the bundle and manifest end up with a requirements.txt file in either case, and with this change, the other case will have to be handled in a code path somewhere else.

Reply:
@mmarchetti thanks for the suggestion.

I changed the handling of the package requirements to be independent of how they are sourced.  One area that was a bit tricky is that the `relevant_files` used to include the `requirements.txt` file, which would cause the bundle to include it twice (once from buffer through `environment.contents`, and another time from `relevant_files`).  With this change, the `requirements.txt` is only added from `environment.contents`.